### PR TITLE
Adding sba client capabilities to allow custom metrics to show up on metrics tab

### PR DESCRIPTION
**Pull request by ctoestreich**
Adding sba client capabilities to allow custom metrics to show up on metrics tab #264  #41

screen shot 2017-05-05 at 9 26 37 am
screen shot 2017-05-05 at 9 26 16 am

screen shot 2017-05-05 at 9 26 24 am
screen shot 2017-05-05 at 9 26 31 am

### CODE REVIEW

**Comment by ctoestreich** [comment]
Perhaps the metrics should be a sorted set instead of a List

File: spring-boot-admin-starter-client/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 1]
```diff
@@ -6,5 +6,11 @@
     "type": "java.lang.Boolean",
     "description": "Enable Spring Admin Client.",
     "defaultValue": "true"
+  },
+  {
+    "name": "spring.boot.admin.client.metrics",
+    "type": "java.util.Set",
```
Comment (ctoestreich) [inline spring-boot-admin-starter-client/src/main/resources/META-INF/additional-spring-configuration-metadata.json]: Technically this is a currently a list...

**Comment by joshiste** [comment]
First of all, thanks for your PR.

I wonder if the solution has to be this complicated? Why don't we just introduce an option in the UI to show all metrics and not only the `gauge.*` and `counter.*` ones.

I think it's a bad idea to extend the `Application` model - your solution won't work for those who use the `DiscoveryClient`. I thinks it's better to put those config into the metadata, which can also be set when using discovery.

**Comment by ctoestreich** [comment]
@joshiste Adding all metrics would be easier.  I think the code there to filter gauge and counters is already a bit overkill.  I don't know what the intent of rich gauges was, but if we made all metrics just simple bars it would at least give cover the use case that we need.  My change has the advantage of allowing clients to dictate which metrics they want to see in the admin.  We have other services that consume all of these metrics and I we emit hundreds of metrics per service.  Rendering a contextual subset is probably suitable to our case, but showing all metrics would encapsulate this as well.  

Your comment has two different suggestions of work.  Show all metrics OR change custom metric set to metadata property.  I am happy to augment to fit one of those, but lets decide which one first and then I will make the change.

**Comment by joshiste** [comment]
I tend to the easier solution and simply show all metrics.

**Comment by ctoestreich** [comment]
Okay i will clean this up and do all metrics grouped by the prefix.  Are we okay to simplify the gauges or do you want to keep those regex tests around and the "rich gauge" feature?  What exactly are the rich gauges supposed to be?

**Comment by joshiste** [comment]
Spring boot has the option to use rich-gauges which provide min/max/avg/... and these are rendered as bars with some additional marks.

**Comment by joshiste** [comment]
When showing all metrics on the page, please consider a search input for filtering the metrics...

**Comment by joshiste** [comment]
@ctoestreich are you still working on this PR?

**Comment by ctoestreich** [comment]
@joshiste After spending a few hours working on this the simplicity of just specifying the interested metrics was pretty nice from a coding and display perspective.  To just boot a microservice below are the number of metrics that are created.  After it gets running there are may more given all the http interceptor, amqp and other custom metrics that start getting counted.  Not to mention our primary monolith has about 2k controller/action that log access rates, counts etc.

So for me to say show: "jvm.memory, web.access, threads" limits this scope to the SBA screen to stuff we want to see at a glance which the bulk of these are emitted to our metrics aggregation platform. 

```json
{

    "mem": 292877,
    "mem.free": 46604,
    "processors": 4,
    "instance.uptime": 81834864,
    "uptime": 37411948,
    "systemload.average": 0.16,
    "heap.committed": 160256,
    "heap.init": 32768,
    "heap.used": 113651,
    "heap": 455168,
    "nonheap.committed": 134720,
    "nonheap.init": 2496,
    "nonheap.used": 132621,
    "nonheap": 0,
    "threads.peak": 50,
    "threads.daemon": 46,
    "threads.totalStarted": 2480,
    "threads": 49,
    "classes": 15061,
    "classes.loaded": 15061,
    "classes.unloaded": 0,
    "gc.ps_scavenge.count": 132,
    "gc.ps_scavenge.time": 1089,
    "gc.ps_marksweep.count": 4,
    "gc.ps_marksweep.time": 509,
    "hystrix.threadPool.IndexController.countThreadsExecuted": 31,
    "hystrix.command.index.latencyExecute_percentile_95": 0,
    "jvm.memory.non-heap.used": 135806624,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.999thPercentile": 141,
    "jvm.classes.unloaded": 0,
    "jvm.memory.total.committed": 302055424,
    "jvm.gc.PS-MarkSweep.time": 509,
    "jvm.threads.blocked.count": 0,
    "counter.status.200.health": 630,
    "jvm.memory.non-heap.committed": 137953280,
    "hystrix.command.index.latencyExecute_percentile_99": 0,
    "jvm.memory.pools.Code-Cache.max": 251658240,
    "jvm.memory.pools.Metaspace.usage": 0.9837533106375774,
    "web.access.controller.Swagger2Controller.action.getDocumentation.oneMinuteRate": 1.1875240736777807E-270,
    "gauge.response.metrics": 7.0,
    "jvm.memory.heap.used": 116494792,
    "jvm.memory.pools.PS-Eden-Space.committed": 45613056,
    "jvm.memory.heap.usage": 0.25004073015348177,
    "jvm.threads.new.count": 0,
    "hystrix.threadPool.IndexController.threadActiveCount": 0,
    "jvm.threads.count": 49,
    "jvm.threads.terminated.count": 0,
    "jvm.memory.pools.PS-Survivor-Space.init": 1048576,
    "jvm.memory.pools.PS-Old-Gen.usage": 0.2586348503604643,
    "jvm.memory.pools.PS-Old-Gen.used": 90444632,
    "jvm.memory.heap.max": 466092032,
    "jvm.memory.heap.committed": 164102144,
    "jvm.memory.non-heap.usage": -1.35804168E8,
    "counter.status.200.root": 32,
    "web.access.controller.Swagger2Controller.action.getDocumentation.count": 1,
    "jvm.fd": 3.62396240234375E-5,
    "counter.status.200.metrics": 2,
    "hystrix.command.index.countTimeout": 0,
    "jvm.memory.pools.Compressed-Class-Space.usage": 0.009296298027038574,
    "web.access.controller.IndexController.action.index.snapshot.stdDev": 2,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.98thPercentile": 141,
    "jvm.memory.pools.Compressed-Class-Space.used": 9981824,
    "jvm.threads.timed_waiting.count": 15,
    "web.access.controller.Swagger2Controller.action.getDocumentation.status.200": 1,
    "web.access.controller.IndexController.action.index.fifteenMinuteRate": 0.0043852199194119295,
    "web.access.controller.IndexController.action.index.oneMinuteRate": 0.03198334378909226,
    "jvm.gc.PS-MarkSweep.count": 4,
    "jvm.memory.pools.PS-Survivor-Space.used": 2737560,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.min": 141,
    "jvm.memory.pools.PS-Eden-Space.init": 8912896,
    "jvm.memory.pools.Compressed-Class-Space.init": 0,
    "web.access.controller.IndexController.action.index.snapshot.max": 19,
    "jvm.memory.pools.PS-Old-Gen.init": 22544384,
    "jvm.threads.deadlock.count": 0,
    "jvm.memory.pools.PS-Eden-Space.max": 159383552,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.99thPercentile": 141,
    "jvm.classes.loaded": 15061,
    "jvm.memory.non-heap.max": -1,
    "jvm.memory.pools.PS-Old-Gen.max": 349700096,
    "jvm.memory.pools.PS-Survivor-Space.max": 6815744,
    "gauge.response.v2.api-docs": 166.0,
    "web.access.controller.IndexController.action.index.snapshot.99thPercentile": 9,
    "web.access.controller.Swagger2Controller.action.getDocumentation.meanRate": 2.687816256783247E-5,
    "web.access.controller.IndexController.action.index.status.200": 32,
    "gauge.response.info": 246.0,
    "hystrix.command.index.countSuccess": 31,
    "jvm.gc.PS-Scavenge.count": 132,
    "hystrix.command.index.countShortCircuited": 0,
    "counter.status.200.v2.api-docs": 1,
    "web.access.controller.IndexController.action.index.snapshot.75thPercentile": 9,
    "jvm.threads.daemon.count": 46,
    "hystrix.command.index.countFailure": 0,
    "jvm.memory.non-heap.init": 2555904,
    "web.access.controller.IndexController.action.index.snapshot.98thPercentile": 9,
    "jvm.gc.PS-Scavenge.time": 1089,
    "jvm.memory.pools.Metaspace.committed": 84672512,
    "jvm.memory.pools.Metaspace.init": 0,
    "web.access.controller.IndexController.action.index.snapshot.999thPercentile": 9,
    "jvm.memory.pools.Metaspace.max": -1,
    "hystrix.command.index.countThreadPoolRejected": 0,
    "web.access.controller.IndexController.action.index.meanRate": 8.560506389129574E-4,
    "jvm.memory.pools.Metaspace.used": 83296864,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.mean": 141,
    "hystrix.threadPool.IndexController.queueSize": 0,
    "jvm.memory.pools.Code-Cache.init": 2555904,
    "hystrix.command.index.latencyExecute_mean": 0,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.stdDev": 0,
    "jvm.memory.pools.Compressed-Class-Space.max": 1073741824,
    "jvm.threads.runnable.count": 7,
    "jvm.memory.pools.Code-Cache.usage": 0.16899083455403646,
    "web.access.controller.Swagger2Controller.action.getDocumentation.fiveMinuteRate": 2.8559623708135745E-55,
    "jvm.memory.pools.PS-Eden-Space.usage": 0.1463085977654708,
    "web.access.controller.IndexController.action.index.snapshot.95thPercentile": 9,
    "jvm.memory.pools.Code-Cache.committed": 43057152,
    "jvm.memory.pools.PS-Survivor-Space.usage": 0.4016524094801683,
    "web.access.controller.IndexController.action.index.count": 32,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.95thPercentile": 141,
    "jvm.memory.pools.Compressed-Class-Space.committed": 10223616,
    "jvm.memory.total.used": 252432120,
    "web.access.controller.Swagger2Controller.action.getDocumentation.fifteenMinuteRate": 2.2521854967661474E-19,
    "jvm.threads.waiting.count": 27,
    "counter.status.200.info": 1,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.median": 141,
    "jvm.memory.pools.PS-Eden-Space.used": 23319184,
    "web.access.controller.IndexController.action.index.snapshot.min": 3,
    "web.access.controller.IndexController.action.index.fiveMinuteRate": 0.007586273947445572,
    "jvm.memory.pools.PS-Survivor-Space.committed": 6815744,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.75thPercentile": 141,
    "jvm.memory.pools.Code-Cache.used": 42527936,
    "web.access.controller.IndexController.action.index.snapshot.median": 9,
    "jvm.memory.total.max": 466092031,
    "web.access.controller.IndexController.action.index.snapshot.mean": 6,
    "gauge.response.health": 18.0,
    "jvm.memory.pools.PS-Old-Gen.committed": 111673344,
    "web.access.controller.Swagger2Controller.action.getDocumentation.snapshot.max": 141,
    "jvm.memory.total.init": 36110336,
    "jvm.memory.heap.init": 33554432,
    "gauge.response.root": 10.0,
    "integration.channel.errorChannel.errorRate.mean": 0.0,
    "integration.channel.errorChannel.errorRate.max": 0.0,
    "integration.channel.errorChannel.errorRate.min": 0.0,
    "integration.channel.errorChannel.errorRate.stdev": 0.0,
    "integration.channel.errorChannel.errorRate.count": 0,
    "integration.channel.errorChannel.sendCount": 0,
    "integration.channel.errorChannel.sendRate.mean": 0.0,
    "integration.channel.errorChannel.sendRate.max": 0.0,
    "integration.channel.errorChannel.sendRate.min": 0.0,
    "integration.channel.errorChannel.sendRate.stdev": 0.0,
    "integration.channel.errorChannel.sendRate.count": 0,
    "integration.channel.nullChannel.errorRate.mean": 0.0,
    "integration.channel.nullChannel.errorRate.max": 0.0,
    "integration.channel.nullChannel.errorRate.min": 0.0,
    "integration.channel.nullChannel.errorRate.stdev": 0.0,
    "integration.channel.nullChannel.errorRate.count": 0,
    "integration.channel.nullChannel.sendCount": 0,
    "integration.channel.nullChannel.sendRate.mean": 0.0,
    "integration.channel.nullChannel.sendRate.max": 0.0,
    "integration.channel.nullChannel.sendRate.min": 0.0,
    "integration.channel.nullChannel.sendRate.stdev": 0.0,
    "integration.channel.nullChannel.sendRate.count": 0,
    "integration.handler._org.springframework.integration.errorLogger.handler.duration.mean": 0.0,
    "integration.handler._org.springframework.integration.errorLogger.handler.duration.max": 0.0,
    "integration.handler._org.springframework.integration.errorLogger.handler.duration.min": 0.0,
    "integration.handler._org.springframework.integration.errorLogger.handler.duration.stdev": 0.0,
    "integration.handler._org.springframework.integration.errorLogger.handler.duration.count": 0,
    "integration.handler._org.springframework.integration.errorLogger.handler.activeCount": 0,
    "integration.handlerCount": 1,
    "integration.channelCount": 2,
    "integration.sourceCount": 0,
    "httpsessions.max": -1,
    "httpsessions.active": 0,
    "datasource.primary.active": 0,
    "datasource.primary.usage": 0.0
}
```

**Comment by ctoestreich** [comment]
I was able to resolve most of the issues I mentioned earlier.  There should be a better way to display timers, but I don't have time for that atm.

screen shot 2017-06-08 at 2 36 42 pm
screen shot 2017-06-08 at 2 36 30 pm
screen shot 2017-06-08 at 2 36 20 pm

**Comment by joshiste** [comment]
looking good so far. I'll have a deeper look next weekend.

**Comment by joshiste** [comment]
Merged with 82d707e6fc3a77b07880d566110c0478265cea71. 
I did modify the filtering to behave like the one on the environment page.
Thank you for contributing to this project!

**Comment by talbert83** [comment]
Hi, I don't seem to find this feature in the metrics tab, using SBA 1.5.4. There's no filter and only counters & gauges are displayed. Do I need to set some configuration? I didn't find any documentation regarding this.

**Comment by joshiste** [comment]
Did you update the ui jar? Did you try clearing your browser cache?

**Comment by talbert83** [comment]
You're right, clearing the cache helped. Thanks for the hint!

**Pull request closed without merging**